// @expo/build-tools/dist/platforms/android/utils/credentials.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.restoreCredentials = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const uuid_1 = require("uuid");
async function restoreCredentials(projectDir, job) {
    if (!job.secrets) {
        throw new Error('secrets are missing in the job object'); // TODO: sentry (should be detected earlier)
    }
    const keystorePath = `keystore-${uuid_1.v4()}.jks`;
    await fs_extra_1.default.writeFile(path_1.default.join(projectDir, keystorePath), Buffer.from(job.secrets.keystore.dataBase64, 'base64'));
    const credentialsJson = {
        android: {
            keystore: {
                keystorePath,
                keystorePassword: job.secrets.keystore.keystorePassword,
                keyAlias: job.secrets.keystore.keyAlias,
                keyPassword: job.secrets.keystore.keyPassword,
            }
        }
    };
    await fs_extra_1.default.writeFile(path_1.default.join(projectDir, 'credentials.json'), JSON.stringify(credentialsJson));
}
exports.restoreCredentials = restoreCredentials;
